pyoomph/c_output.py

```diff
@@ -22,224 +22,186 @@
-
-        if isinstance(ast, (tast.ReturningCall, tast.VoidCall)):
-            return self.emit_call(ast) + ";\n\t"
-
-        if isinstance(ast, tast.DecRef):
-            return self.file_emitter.emit_decref(
-                self.emit_expression(ast.value), ast.value.type
+        if isinstance(ins, tast.CallConstructor):
+            return self.emit_call(
+                "ctor_" + self.file_emitter.get_type_c_name(ins.result.type),
+                ins.args,
+                ins.result,
             )
-
-        if isinstance(ast, tast.Return):
-            if ast.value is not None:
-                return f"""
-                retval = {self.emit_expression(ast.value)};
-                {self.file_emitter.emit_incref("retval", ast.value.type)}
-                goto out;
-                """
+        if isinstance(ins, tast.Return):
+            if ins.value is not None:
+                return f"{self.incref_var(ins.value)}; retval = {self.emit_local_var(ins.value)}; goto out;"
             return "goto out;"
-
-        if isinstance(ast, tast.If):
+        if isinstance(ins, tast.GetAttribute):
+            return f"{self.emit_local_var(ins.result)} = {self.emit_local_var(ins.obj)}->memb_{ins.attribute}; {self.incref_var(ins.result)};"
+        if isinstance(ins, tast.PointersEqual):
+            return f"{self.emit_local_var(ins.result)} = ({self.emit_local_var(ins.lhs)} == {self.emit_local_var(ins.rhs)});"
+        if isinstance(ins, tast.If):
+            then = "\n\t\t".join(map(self.emit_instruction, ins.then))
+            otherwise = "\n\t\t".join(map(self.emit_instruction, ins.otherwise))
             return f"""
-            if ({self.emit_expression(ast.condition)}) {{
-                {"".join(self.emit_statement(s) for s in ast.then)}
+            if ({self.emit_local_var(ins.condition)}) {{
+                {then}
             }} else {{
-                {"".join(self.emit_statement(s) for s in ast.otherwise)}
+                {otherwise}
             }}
             """
-
-        if isinstance(ast, tast.Loop):
+        if isinstance(ins, tast.Loop):
             # While loop because I couldn't get C's for loop to work here
+            cond_code = "\n\t\t".join(map(self.emit_instruction, ins.cond_code))
+            body = "\n\t\t".join(map(self.emit_instruction, ins.body))
+            incr = "\n\t\t".join(map(self.emit_instruction, ins.incr))
             return f"""
-            {"".join(self.emit_statement(s) for s in ast.init)}
-            while ({self.emit_expression(ast.cond)}) {{
-                {"".join(self.emit_statement(s) for s in ast.body)}
-                {self.emit_label(ast.loop_id)}  // oomph 'continue' jumps here
-                {"".join(self.emit_statement(s) for s in ast.incr)}
+            while(1) {{
+                {cond_code}
+                if (!{self.emit_local_var(ins.cond)})
+                    break;
+                {body}
```

Comment:
Perhaps this pattern should be exposed in the language itself? Arbitrary stuff in beginning and end of loop?